extensions/sql-migration/src/models/loginMigrationModel.ts

```diff
@@ -0,0 +1,247 @@
+				this._markLoginStatus(loginName, loginStatus);
+			}
+		}
+	}
+
+	public ReportException(step: LoginMigrationStep, error: any): void {
+		this._currentStepIdx = this._loginMigrationSteps.findIndex(s => s === step) + 1;
+
+		for (const loginName of this._logins.keys()) {
+			// Mark current step as failed with the error message and mark remaining messages as canceled
+			let errors = [error.message];
+			this._addStepStateForLogin(loginName, step, LoginMigrationState.Failed, errors);
+			this._markRemainingSteps(loginName, LoginMigrationState.Canceled);
+			this._markLoginStatus(loginName, LoginMigrationState.Failed);
+		}
+
+		this._markMigrationComplete();
+	}
+
+	public GetLoginMigrationResults(loginName: string): MultiStepResult[] {
+		let loginResults: MultiStepResult[] = [];
+		let login = this._getLogin(loginName);
+
+		for (const step of this._loginMigrationSteps) {
+			// The default steps and state will be added if no steps have completed
+			let stepResult: MultiStepResult = {
+				stepName: GetLoginMigrationStepString(step),
+				state: MultiStepState.Pending,
+				errors: [],
+			}
+
+			// If the step has completed, then the login will have the stored status
+			if (login?.statusPerStep.has(step)) {
+				let stepStatus = login!.statusPerStep.get(step);
+				stepResult.state = LoginMigrationStateToMultiStepState[stepStatus!.status];
+				stepResult.errors = stepStatus!.errors;
+			} else if (step === this.currentStep) {
+				stepResult.state = LoginMigrationStateToMultiStepState[LoginMigrationState.Running];
+			}
+
+			loginResults.push(stepResult);
+		}
+
+		return loginResults;
+	}
+
+	public AddNewLogins(logins: string[]) {
+		logins.forEach(login => this._addNewLogin(login));
+	}
+
+	public SetLoginMigrationSteps(steps: LoginMigrationStep[] = []) {
+		this._loginMigrationSteps = [];
+
+		if (steps.length === 0) {
+			this._loginMigrationSteps.push(LoginMigrationStep.MigrateLogins);
+			this._loginMigrationSteps.push(LoginMigrationStep.EstablishUserMapping);
+			this._loginMigrationSteps.push(LoginMigrationStep.MigrateServerRolesAndSetPermissions);
+		} else {
+			this._loginMigrationSteps = steps;
+		}
+	}
+
+
+	private _getLogin(loginName: string) {
+		return this._logins.get(loginName.toLocaleLowerCase());
+	}
+
+	private _addNewLogin(loginName: string, status: LoginMigrationState = LoginMigrationState.Pending) {
+		let newLogin: Login = {
+			loginName: loginName,
+			overallStatus: status,
+			statusPerStep: new Map<LoginMigrationStep, LoginMigrationStepState>(),
+		}
+
+		this._logins.set(loginName.toLocaleLowerCase(), newLogin);
+	}
+
+	private _addStepStateForLogin(loginName: string, step: LoginMigrationStep, stepStatus: LoginMigrationState, errors: string[] = []) {
+		const loginExist = this._logins.has(loginName);
+
+		if (!loginExist) {
+			this._addNewLogin(loginName, LoginMigrationState.Running);
+		}
+
+		let login = this._getLogin(loginName);
```

Comment:
nit, use const here.

Reply:
hmm, so this login changes, in terms of its member variables, which is why I uses let. but i guess const only applies for assignment and not the underlying member variables. made the change!